docs/templates/docker-in-docker.md

```diff
@@ -0,0 +1,170 @@
+
+```hcl
+resource "coder_agent" "main" {
+  os   = "linux"
+  arch = "amd64"
+  dir  = "/home/coder"
+  startup_script = <<EOF
+    #!/bin/sh
+
+    # Start Docker
+    sudo dockerd &
+
+    # ...
+  EOF
+}
+
+resource "k8s_core_v1_pod" "dev" {
+  count = data.coder_workspace.me.start_count
+  metadata {
+    name      = "coder-${data.coder_workspace.me.owner}-${data.coder_workspace.me.name}"
+    namespace = var.workspaces_namespace
+    annotations = {
+      "io.kubernetes.cri-o.userns-mode" = "auto:size=65536"
+    }
+  }
+
+  # Use the Sysbox container runtime (required)
+  runtime_class_name = "sysbox-runc
+
+  spec {
+    security_context {
+      run_asuser = 1000
+      fsgroup    = 1000
+    }
+    containers {
+      name = "dev"
+      env {
+        name  = "CODER_AGENT_TOKEN"
+        value = coder_agent.main.token
+      }
+      image = "codercom/enterprise-base:ubuntu"
+      command = ["sh", "-c", coder_agent.main.init_script]
+    }
+  }
+}
+```
+
+> Sysbox CE (Community Edition) supports a maximum of 16 pods (workspaces) per node on Kubernetes. See the [Sysbox documentation](https://github.com/nestybox/sysbox/blob/master/docs/user-guide/install-k8s.md#limitations) for more details.
+
+## Privileged sidecar container (Kubernetes)
+
+While less secure, you can attach a [privileged container](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) to your templates. This may come in handy if your nodes cannot run Sysbox.
+
+```hcl
+resource "coder_agent" "main" {
+  os             = "linux"
+  arch           = "amd64"
+}
+
+resource "kubernetes_pod" "main" {
+  count = data.coder_workspace.me.start_count
+  metadata {
+    name      = "coder-${data.coder_workspace.me.owner}-${data.coder_workspace.me.name}"
+    namespace = var.namespace
+  }
+  spec {
+    # Run a privileged dind (Docker in Docker) container
+    container {
+      name  = "docker-sidecar"
+      image = "docker:dind"
+      security_context {
+        privileged = true
+      }
+      command = ["dockerd", "-H", "tcp://127.0.0.1:2375"]
+    }
+    container {
+      name    = "dev"
+      image   = "codercom/enterprise-base:ubuntu"
+      command = ["sh", "-c", coder_agent.main.init_script]
+      security_context {
+        run_as_user = "1000"
+      }
+      env {
+        name  = "CODER_AGENT_TOKEN"
+        value = coder_agent.main.token
+      }
+      # Use the Docker daemon in the "docker-sidecar" container
+      env {
+        name  = "DOCKER_HOST"
+        value = "localhost:2375"
+      }
+    }
+  }
+}
+
+```
+
+## Shared Docker socket (Docker)
+
+While less secure, Docker-based templates can share the host's Docker socket.
+
+````hcl
+resource "coder_agent" "main" {
+  arch           = data.coder_provisioner.me.arch
+  os             = "linux"
+  startup_script = <<EOF
+    #!/bin/sh
+
+    # Give the internal "coder" user permission
+    # to use the Docker socket
+    sudo chmod 666 /var/run/socker.sock
```

Comment:
I need some help here since I'm assuming this is the wrong thing to do.

I got kinda stuck figuring out how to add the inner `coder` user to the shared socket. Wondering if I should just add a `dind` approach for Docker hosts (2 docker_container resources and a network) and remove the "shared daemon" use case entirely?

Reply:
In this approach, all Coder workspaces have unfettered access to each other. It's pretty dangerous, so perhaps we merge this w/o the privileged approach and we revisit when requested.

Reply:
Makes sense, I added a privileged sidecar method for docker.